content/backend/graphql-scala/4-deferred_resolvers.md

```diff
@@ -0,0 +1,168 @@
+
+ok, if we have our fetcher defined, we push it to lower level.
+
+<Instruction>
+
+in `GraphQLSchema` create constant for deferred resolver.
+
+```scala
+
+val Resolver = DeferredResolver.fetchers(linksFetcher)
+
+```
+
+</Instruction>
+
+Such resolver have to be passed into the `Executor` to make it available for use. As you can see, we use the same fetcher in two fields, in the first example we're providing only single id and expecting one, optional object (`deferOpt` function). In the second case we're providing a list of ids and expecting a sequence of objects (`deferSeq`).
+
+<Instruction>
+
+Add resolver to the executor, open the `GraphQLServer.scala` file, and change `executeGraphQLQuery` function content as follows:
+
+```scala
+
+Executor.execute(
+  GraphQLSchema.SchemaDefinition,
+  query,
+  MyContext(dao),
+  variables = vars,
+  operationName = operation,
+  deferredResolver = GraphQLSchema.Resolver
+).map( //the rest without changes
+
+```
+
+</Instruction>
+
+Since, we're using `DAO.getLinks` to fetch single entity or entire list, we don't need `getLink` function anymore.
+
+<Instruction>
+
+Open a `DAO` class and remove `getLink` function.
+
+</Instruction>
+
+
+### HasId type class
+
+If you tried to execute a query, you get an error. Because there one more thing we have to do. `Fetcher` needs 'something' what extraxts id from entity. This thing is `HasId` type class. You have few choices how to provide usch class for your model. Firstly you can explicitly pass it, like this:
+
+```scala
+
+val linksFetcher = Fetcher(
+  (ctx: MyContext, ids: Seq[Int]) => ctx.dao.getLinks(ids)
+)(HasId(_.id))
+
+```
+
+On the other hand. You can declare implicit constant in the same context so fetcher will take it implicitly.
+The third way is to provide `HasId` implicitly inside companion object for our model, like this:
+
+```scala
+
+object Link {
+    implicit val hasId = HasId[Link, Int](_.id)
+}
+
+```
+There one more way, more generic, but to avoid a mess, I'll back to this in chapter about interfaces.
+
+<Instruction>
+
+Add implicit hasId in the `GraphQLServer`
+
+```scala
+
+implicit val linkHasId = HasId[Link, Int](_.id)
+
+```
+
+</Instruction>
+
+
+### Test it
+
+You can debug `DAO.getLinks` function in any way, and execute following query:
+
+```graphql
+
+query {
+
+    l1: link(id: 1){
+    	id
+    	url
+  	}
+
+  l2: link(id: 1){
+    	id
+    	url
+  	}
+
+  	links(ids: [2,3]){
+      id
+      url
+    }
+
+}
+
+```
+
+As you can see `getLinks` is called only once!
+
+### What's next?
+
+In the next chapter we will extern `Link` model about date fields. H2 doesn't support the type we need so we have to manage it somehow. You will learn about custom scalar types.
```

Comment:
This would need more "glue" to encourage readers to carry on ... but I don't have any ideas atm :/